### DISCUSSION TITLE
Min/default/max for custom blocks

### DISCUSSION THREAD

**Discussion by WarpedWartWars** [WarpedWartWars/Snap-testing#3] [Ideas]
I'm hoping to make Min/Default/Max for custom blocks. No promises as to if it'll work.

(@brianharvey this was your idea)

**Comment by WarpedWartWars** [comment]
What monstrosity did I make?
![Screenshot (214)](https://user-images.githubusercontent.com/91039320/141022342-a940affe-af98-4e7e-bec7-e833d5c2d568.png)

**Reply by brianharvey** [reply]
That's pretty awful, but I do like the upvar picture.  What are you after with the solid grey inline command?
And of course multiple inputs goes at the end.

**Reply by WarpedWartWars** [reply]
Yeah.
Just something I wanted.
Yeah... I'm going to fix that eventually. I'm currently getting it so min/default/max actually works.

**Reply by DarDoro** [reply]
MultArgMorph has a _"minInputs"_ attribute that sets the number of slots created at initialization. "max/min inputs" can be handled at fixSlotsLayout.
But how do you want to store those values at project save / load?
Do you want to extend v2 format to be incompatible with Snap.7? Or maybe store parameters in a _"default value"_ placeholder or encoded as part of parameter name (e.g: myList:2,0 ), or maybe parsed from comment attached to the block definition header?

BTW: number of parameter types is hardcoded in _InputSlotDialogMorph.prototype.fixSlotsLayout_
`size = 12, // number slot type radio buttons`

My example of _color picker_ input type 
https://snap.berkeley.edu/snap/snap.html#present:Username=dardoro&ProjectName=colorPicker&editMode&noRun

**Reply by WarpedWartWars** [reply]
Yeah, I knew that.

IDK.

I don't really care that much with being compatible with v7.

I already set that to 14.

I might add that eventually.

Here's what it looks like now:
![Screenshot (215)](https://user-images.githubusercontent.com/91039320/141034693-95c4afc8-6b18-409b-a617-859bc4764536.png)

**Reply by brianharvey** [reply]
> I don't really care that much with being compatible with v7.

That's silly, if you want your work actually to be useful.

Also, the arrangement of input types in the dialog isn't arbitrary; look in the manual for the discussion of what the rows and columns mean.  You should find a way not to ruin that if you want to add types.  So your BYOB inline command type (which I still don't get why you want) should be the 13th type, at the bottom of the left column.

You're right that "upvar" is mutually exclusive with all the type radio buttons, so could be up there, but it's (so far anyway) incompatible with multiple inputs.  You need a reasonable definition of the syntax (how do you make unique names out of the name that the user provides) and the semantics (what does the program see for this variable).

There's no reason for those super long text boxes for min/def/max.  Min and default will almost always be ≤ 2; max will be either ≤ 3 or infinite (represented by an empty box or a zero).

**Reply by WarpedWartWars** [reply]
> Also the arrangement...

[Here.](https://github.com/WarpedWartWars/Snap-testing/commit/03dab11c91eaf269da3921d0e7a0c7fe5d979896)
> (which I still don't get why you want)

And I still don't know either. I just know I want the three BYOB grey ring slots.
> You're right that "upvar"...

Yeah... I'm not sure what I'll do about that. (maybe make ITEM N OF able to be dropped into SET)
> There's no reason for those...

Yeah! I tried to have it set the width to 15 instead of 50, but it didn't work.
> (represented by an empty box or a zero)

An empty box might not be that intuitive. I already have it so that 0 makes it infinite.

**Comment by WarpedWartWars** [comment]
https://github.com/WarpedWartWars/Snap-testing/blob/3768fc19a96f69631e6183f20e713203e2e9402a/src/byob.js#L3302
How do I edit that so that the BoxMorph is the right size? Pic of what it currently is:
![Screenshot (216)](https://user-images.githubusercontent.com/91039320/141189009-3bc25335-9372-4cb4-8494-08b94b45726b.png)

**Comment by DarDoro** [comment]
You can control the width of inputs with
`minInput.contents().minWidth =30;`

**Reply by WarpedWartWars** [reply]
Thanks! That worked.
